qdev_wrappers/qdev_fitter.py

```diff
@@ -0,0 +1,141 @@
+import qcodes as qc
+import numpy as np
+import types
+from os.path import sep
+import matplotlib.pyplot as plt
+import scipy.fftpack as fftpack
+from scipy.optimize import curve_fit
+
+from qdev_wrappers.file_setup import CURRENT_EXPERIMENT
+
+
+
+
+def qdev_fitter(id, fitfunction_in, do_plots=True, dataname=None, samplefolder=None, p0=None,**kwargs):
+    '''
+    Function to allow easy on the fly fitting of data. Only works for 1D datasets at the moment.
+    id (num): Id of dataset
+    fitfunction_in (str, function): Function data will be fitted to. Can be string for functions defined here or a function passed direclty defined elsewhere.
+    do_plots (bool): If plots are done. Will be saved in the analysis folder for the samplepath
+    dataname (str): String with dataname in the qcodes dataset
+    samplefolder (path): Path to samplefolder if not opening active measurements
+    p0 (tuple): Initial parameters for fit
+    kwargs: Passed to scipy.optimize.curve_fit when fitting.
+    '''
+
+    # Load data from active samplefolder or input samplefolder.
+    str_id = '{0:03d}'.format(id)
+    if samplefolder==None:
+        check_experiment_is_initialized()
+
+        path = qc.DataSet.location_provider.fmt.format(counter=str_id)
+        data = qc.load_data(path)
+    else:
+        path = '{}{}{}'.format(samplefolder,sep,str_id)
+        data = qc.load_data(path)
+
+    # Get dataname to be fitted
+    if dataname==None:
+        # Function only made for a single dataname. Check there isnt more datanames in the dataset.
+        key_data = [key for key in data.arrays.keys() if "_set" not in key[-4:]]
+        if len(key_data) != 1:
+            raise RuntimeError('Dataset has more than one dataname. Choose dataname to be fitted from: {}.'.format(', '.join(key_data)))
+    else:
+        # Check input dataname is in dataset.
+        if dataname not in [key for key in data.arrays.keys() if "_set" not in key[-4:]]:
+            raise RuntimeError('Dataname not in dataset. Input dataname was: \'{}\' while dataname(s) in dataset are: {}.'.format(dataname,', '.join(key_data)))
+        key_data = [dataname]
+
```

Comment:
Here I would check if fitfunction is a `callable`, so you can also pass a class to it for more complicated fit routines. 
Furthermore I am not sure whether it makes a lot of sense to have a string describing the function. What do you have in mind? Is it not that the only difference is the quotation marks?

Reply:
I want an easy way to acces functions defined within the file (T1,T2) instead of having to pass a function.
Most often we are using a same few functions over and over. However, I also want to be able to pass a function for that one odd time where we need to fit something different.

Reply:
I am not sure what exactly you mean. You can do exactly what you described if you define the function in the beginning of the file:
```
def fitfun(x,p1):
   return sin(p1*x)

def dothefit(fitfun, parmeters, constraints):
   ...
```

Reply:
I have fitfun defined as in your example. Now I want to run dothefit() from outside the file where fitfun is not known (I dont want to import all functions in the file).

Reply:
ok, now I get, what you meant was inconvenient and why you chose to circumvent it like this. I think there are two options you might find interesting to consider:
 - you can wrap everything in a class and define your fit function as a static method
 - you can make a module, so that you, just you would do with numpy, import the whole module and then use the dot notation. eg 
```
import ....thorvaldsFit as fit
from fit import dothefit
...
dothefit(fit.fitfun,parameters, constraints,...)
```
- a third option is to do a star import, which, as you know is not considered best practice, but which is, to my mind, a better solution than the workaround with the string for the name of the function.

Reply:
I dont like the start import but the other two options are interesting. Will take a look at them.